### PR TITLE
Unify CPU hist sketching

### PR DESCRIPTION
**Pull request by trivialfis**
@SmirnovEgorRu @ShvetsKS

This PR merges sparse cut builder and dense cut builder into a single sketch container.  At the same time it removes duplicated set of tests and adds tests for distributed environment.  I haven't optimize sparse dataset on distributed environment yet, so it will still generate OOM error as before.  But for single node, as after this PR we don't need to generate an extra copy of dataset from `GetBatches<CSCPage>()`, memory usage should be quite good for sketching.

- Drive by optimization:  I avoided the call to `operator[]` of `SparsePage` by introducing a data view.  So we won't be needing to call `rabit::IsDistributed()` and `HostVector` at every row access.  If it's a burden for review I can revert it.

## Motivation

I'm trying to add categorical data support in XGBoost.  The first hurdle is sketching.  Currently XGBoost has 7 entry points for sketching, I'm trying to merge as many of them as possible.

## Performance for sketching

* Single node 1 x 16

** Before
*** URL
GmatInitialization: 9.75871s, 1 calls @ 9758709us
GmatInitialization: 9.71126s, 1 calls @ 9711261us
GmatInitialization: 9.80294s, 1 calls @ 9802938us

*** HIGGS
GmatInitialization: 3.74729s, 1 calls @ 3747286us
GmatInitialization: 3.90515s, 1 calls @ 3905151us
GmatInitialization: 3.79345s, 1 calls @ 3793448us

** After
*** URL
GmatInitialization: 10.3033s, 1 calls @ 10303323us
GmatInitialization: 10.2406s, 1 calls @ 10240576us
GmatInitialization: 10.224s, 1 calls @ 10224020us

*** HIGGS
GmatInitialization: 3.74322s, 1 calls @ 3743224us
GmatInitialization: 3.63079s, 1 calls @ 3630793us
GmatInitialization: 3.67905s, 1 calls @ 3679049us

* 4 x 4

*** HIGGS
** Before
GmatInitialization: 3.83843s, 1 calls @ 3838427us
GmatInitialization: 3.89644s, 1 calls @ 3896437us
GmatInitialization: 3.90035s, 1 calls @ 3900354us

** After
GmatInitialization: 3.7682s, 1 calls @ 3768198us
GmatInitialization: 3.707s, 1 calls @ 3707000us
GmatInitialization: 3.70828s, 1 calls @ 3708276us

URL for distributed is not tested as it will blow up the memory.  I have another branch for optimizing it, the target is we can unify the sketching between `hist` and `approx`.

### CODE REVIEW

**Comment by trivialfis** [comment]
Will reopen when ready.

**Comment by trivialfis** [comment]
Interestingly this PR exposes some issues in handling empty column via ubsan.

**Comment by trivialfis** [comment]
Maybe dask is leaking file descriptor.

**Comment by trivialfis** [comment]
Dropped some optimizations to simplify the code, mostly affecting sparse datasets like URL.  Will do a new bench but I think the impact should be small.

File: include/xgboost/data.h [hunk 1]
```diff
@@ -270,6 +285,11 @@ class SparsePage {
             static_cast<Inst::index_type>(size)};
   }
 
+  HostSparsePageView GetView() const {
```
Comment (RAMitchell) [inline include/xgboost/data.h:288]: Is there a way to fix operator[] instead of creating a whole new class?

Comment (trivialfis) [reply include/xgboost/data.h:288]: @RAMitchell If you want to call `operator[]` without first calling `HostVector` twice, then I don't think so.

**Comment by trivialfis** [comment]
## Before
- URL
Build: 3.18824s, 1 calls @ 3188235us
Build: 3.24613s, 1 calls @ 3246126us
Build: 3.29904s, 1 calls @ 3299044us

Mean: 3.244469

- HIGGS
Build: 2.53432s, 1 calls @ 2534319us
Build: 2.5671s, 1 calls @ 2567101us
Build: 2.4364s, 1 calls @ 2436397us

Mean: 2.512606

## After
- URL
SketchOnDMatrix: 3.47337s, 1 calls @ 3473373us
SketchOnDMatrix: 3.36139s, 1 calls @ 3661388us
SketchOnDMatrix: 3.56527s, 1 calls @ 3565270us

Mean: 3.466676

- HIGGS
SketchOnDMatrix: 2.48182s, 1 calls @ 2481822us
SketchOnDMatrix: 2.40048s, 1 calls @ 2400475us
SketchOnDMatrix: 2.43417s, 1 calls @ 2434170us

Mean: 2.438823

So slightly faster on dense dataset and slightly slower on extreme sparse dataset.  I variance should be negligible compared to total training time, with the benefits that now all tests run through the pipeline and easier to add categorical data support.

**Comment by trivialfis** [comment]
@SmirnovEgorRu Please help taking a look.

**Comment by ShvetsKS** [comment]
@trivialfis sorry for long response.
Could you also run distributed version with dmlc-submit on single node.

Initially before unification we observed slowdown of 'GmatInitialization' step (https://github.com/dmlc/xgboost/blob/d268a2a4631077c21562b606680c94a743ab7b18/src/tree/updater_quantile_hist.cc#L86) 
with configurations 1x56 (single worker with 56 available  threads) vs 7x8 (7 workers with 8 available threads): 19.37s vs 5.7s. 

It would be good to be sure that no regression was introduced for distributed case also.

**Comment by trivialfis** [comment]
Let me find a machine with that many threads.

> 19.37s vs 5.7s.

You mean after this PR the code slowed down for 4 times?

**Comment by trivialfis** [comment]
@ShvetsKS Speaking of distributed training, as we are merging rabit into XGBoost https://github.com/dmlc/xgboost/issues/5995 .  I think you can choose other backends in the future like UCX for better CPU perf if model recovery is not of concern.  Feel free to join the discussion and development.

**Comment by ShvetsKS** [comment]
Seems it could be even reproduced on similar configuration: 1x24 vs 4x6 for example.
Described slowdown was observed without this PR. I think performance check for distributed case is also needed to be sure that no other additional regression was introduced.

**Comment by trivialfis** [comment]
@ShvetsKS I did a quick run with dmlc-submit on HIGGS with a 4x4 configuration.

```
* With submit
** Before
GmatInitialization: 3.96191s, 1 calls @ 3961914us
GmatInitialization: 4.06724s, 1 calls @ 4067237us
GmatInitialization: 4.04202s, 1 calls @ 4042019us

** After
GmatInitialization: 4.0002s, 1 calls @ 4200197us
GmatInitialization: 4.06724s, 1 calls @ 4067243us
GmatInitialization: 3.98779s, 1 calls @ 3987792us
```

If there's slow down on GMatInit, I don't think it comes from sketching as I'm the last guy who made changes to it, and it has been a while now.

**Comment by trivialfis** [comment]
Result with setting `nthread` to 4 explicitly in above quick run script, value come from rank 0:

```
** Before

GmatInitialization: 3.83843s, 1 calls @ 3838427us
GmatInitialization: 3.89644s, 1 calls @ 3896437us
GmatInitialization: 3.90035s, 1 calls @ 3900354us

** After
```

**Comment by trivialfis** [comment]
@ShvetsKS  I summarized the perf I got so far.  Could you please take a look?

**Comment by ShvetsKS** [comment]
@trivialfis thanks a lot for information providing.
When we considered 1x56 or 1x16 configuration I mean this configuration:
./dmlc-core/tracker/dmlc-submit --cluster=local --num-workers=1 --worker-cores=16 python higgs-perf.py

I don't have any other concerns, look good generally.

**Comment by trivialfis** [comment]
@ShvetsKS I think the perf in description page should show the perf of single node vs multi node.

**Comment by trivialfis** [comment]
Also thanks for the explanation.

**Comment by trivialfis** [comment]
@ShvetsKS  Could you please explicitly approve the PR?  In general I think it would be best for me to have confirmation from you guys when I make big change to CPU code path.

**Comment by trivialfis** [comment]
@SmirnovEgorRu @ShvetsKS Thanks for the review.

**Pull request merged**